libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java

```diff
@@ -0,0 +1,157 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.ssl;
+
+import org.elasticsearch.common.Nullable;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedKeyManager;
+import javax.net.ssl.X509ExtendedTrustManager;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.util.Collection;
+import java.util.Locale;
+
+/**
+ * A variety of utility methods for working with or constructing {@link KeyStore} instances.
+ */
+final class KeyStoreUtil {
+
+    private KeyStoreUtil() {
+        throw new IllegalStateException("Utility class should not be instantiated");
+    }
+
+    /**
+     * Make a best guess about the "type" (see {@link KeyStore#getType()}) of the keystore file located at the given {@code Path}.
+     * This method only references the <em>file name</em> of the keystore, it does not look at its contents.
+     */
+    static String inferKeyStoreType(Path path) {
+        String name = path == null ? "" : path.toString().toLowerCase(Locale.ROOT);
+        if (name.endsWith(".p12") || name.endsWith(".pfx") || name.endsWith(".pkcs12")) {
+            return "PKCS12";
+        } else {
+            return "jks";
+        }
+    }
+
+    /**
+     * Read the given keystore file.
+     *
+     * @throws SslConfigException       If there is a problem reading from the provided path
+     * @throws GeneralSecurityException If there is a problem with the keystore contents
+     */
+    static KeyStore readKeyStore(Path path, String type, char[] password) throws GeneralSecurityException {
+        try {
+            KeyStore keyStore = KeyStore.getInstance(type);
+            try (InputStream in = Files.newInputStream(path)) {
+                keyStore.load(in, password);
+            }
+            return keyStore;
+        } catch (FileNotFoundException | NoSuchFileException e) {
```

Comment:
I know this code was mainly lifted, but I think it might be nicer to have

```java
if (Files.exists(path)) {
// attempt to load
} else {
    throw new SslConfigException("...");
}
```

within the try block and only catch `IOException`.